File: scripts/migrate.js
const Enmap = require('enmap');
const sql = require('sqlite3');
const mongodb = require('mongodb');
const usersettingskeys = require('../usersettings-keys.json');
const guildsettingskeys = require('../guildsettings-keys.json');

function migrate() {
	process.stdout.write('This process may take a while, press any to start the process. Do not stop it before it finishes!');
	process.stdin.setRawMode(true);

	process.stdin.once('data', data => {
		process.stdin.setRawMode(false);

		const byteArray = [...data];
		if (byteArray.length > 0 && byteArray[0] === 3) {
			console.log('^C');
			process.exit(1);
		}

		process.stdout.clearLine();
		process.stdout.cursorTo(0);
		process.stdout.write('1/3 Loading db botconfs...');
		const botconfs = new Enmap({
			name: 'botconfs',
			fetchAll: true
		});

		botconfs.defer.then(() => {
			process.stdout.clearLine();
			process.stdout.cursorTo(0);
			process.stdout.write('1/3 Loading db userdb...');
			const userdb = new Enmap({
				name: 'userdb',
				fetchAll: true
			});
			userdb.defer.then(() => {
				process.stdout.clearLine();
				process.stdout.cursorTo(0);
				process.stdout.write('1/3 Loading db guildsettings...');
				const guildconfs = new Enmap({
					name: 'guildsettings',
					fetchAll: true
				});
				

				guildconfs.defer.then(async () => {
					const settingsFile = require('../settings.json');
					this.url = `mongodb://${encodeURIComponent(settingsFile.db.user)}:${encodeURIComponent(settingsFile.db.password)}@${encodeURIComponent(settingsFile.db.host)}:${encodeURIComponent(settingsFile.db.port)}/?authMechanism=DEFAULT&authSource=admin`;
					try {
						this.dbClient = await mongodb.MongoClient.connect(this.url, { useNewUrlParser: true });
					} catch (err) {
						console.log(err);
						process.exit(-1);
					}

					this.db = this.dbClient.db('lenoxbot');
					const guildSettingsCollection = this.db.collection('guildSettings');
					const userSettingsCollection = this.db.collection('userSettings');
					const botSettingsCollection = this.db.collection('botSettings');

					await guildSettingsCollection.createIndex('guildId', { unique: true });
					await userSettingsCollection.createIndex('userId', { unique: true });
					await botSettingsCollection.createIndex('botconfs', { unique: true });

					process.stdout.clearLine();
					process.stdout.cursorTo(0);
					process.stdout.write('2/3 Converting botconfs...');

					let settings = {};

					for (var [key, value] of botconfs) {
						settings[key] = value;
					}
					for (var key2 in settings['botconfs']) {
						settings[key2] = settings['botconfs'][key2];
					}
					delete settings['botconfs']

					await botSettingsCollection.insertOne({ botconfs: 'botconfs', settings: settings });

					settings = {};

					process.stdout.clearLine();
					process.stdout.cursorTo(0);
					process.stdout.write('2/3 Converting guildSettings...');

					for (var [key, value] of guildconfs) {
						await guildSettingsCollection.insertOne({ guildId: key, settings: value });
					}

					settings = {};

					process.stdout.clearLine();
					process.stdout.cursorTo(0);
					process.stdout.write('2/3 Converting userdb...');

					for (var [key, value] of userdb) {
						await userSettingsCollection.insertOne({ userId: key, settings: value });
					}

					process.stdout.clearLine();
					process.stdout.cursorTo(0);
					process.stdout.write('3/3 Finalizing...');

					this.dbClient.close();
				});
			});
		});
	});
}

migrate();

File: scripts/migrate.py
from pymongo import MongoClient
import json
import sqlite3
from colorama import init
import os
import sys

if os.name == 'nt':
    init()

userSettingsKeys = json.load(open("usersettings-keys.json", "r"))
guildSettingsKeys = json.load(open("guildsettings-keys.json", "r"))

print("Starting conversion...")

client = MongoClient('mongodb://user:password@localhost:27017/')
db = client.get_database(name="lenoxbot")
userSettingsCollection = db.get_collection(name="userSettings")
guildSettingsCollection = db.get_collection(name="guildSettings")
conn = sqlite3.connect('lenoxbotscore.sqlite')
c = conn.cursor()
c.execute("SELECT * FROM medals")
count = 0
for row in c:
    settings = {}
    doc = userSettingsCollection.find_one({"userId": row[0]})
    if row[1] == 0:
        count = count + 1
        continue
    if doc == None:
        settings = userSettingsKeys
        doc = {}
        doc["userId"] = row[0]
        doc["settings"] = settings
        userSettingsCollection.insert_one(doc)
    else:
        settings = doc["settings"]
    settings["credits"] = row[1]
    userSettingsCollection.update_one({"userId": row[0]}, {"$set": {"settings": settings}})
    count = count + 1
    if (count % 100) == 0:
        sys.stdout.write("Inserting user number {0}\n".format(count))
        sys.stdout.write("\033[1A")
        sys.stdout.flush()
print("Inserted %d documents" % count)
count = 1
countUser = 0
guilds = []
c.execute("SELECT * FROM scores")
for row in c:
    settings = {}
    doc = guildSettingsCollection.find_one({"guildId": row[0]})

    if row[0] in guilds:
        continue

    if doc == None:
        settings = guildSettingsKeys
        doc = {}
        doc["guildId"] = row[0]
        doc["settings"] = settings
        guildSettingsCollection.insert_one(doc)
    else:
        settings = doc["settings"]
    if not "scores" in settings:
        settings["scores"] = {}
    curInner = conn.cursor()
    curInner.execute("SELECT * FROM scores WHERE guildId={0}".format(row[0]))
    if (count % 100) == 0:
        sys.stdout.write("Inserting guild #{0}\n".format(count))
        sys.stdout.write("\033[1A")
        sys.stdout.flush()
    for rowInner in curInner:
        settings["scores"][rowInner[1]] = {}
        settings["scores"][rowInner[1]]["points"] = rowInner[2]
        settings["scores"][rowInner[1]]["level"] = rowInner[3]
        countUser = countUser + 1
        sys.stdout.write("\033[2K")
        sys.stdout.write("Inserting guild #{0} > user #{1}\n".format(count, countUser))
        sys.stdout.write("\033[1A")
        sys.stdout.flush()
    guildSettingsCollection.update_one({"guildId": row[0]}, {"$set": {"settings": settings}})
    countUser = 0
    count = count + 1
    guilds.append(row[0])
print("Inserted %d documents" % count)
exit()